pkg/services/auth/auth.go

```diff
@@ -0,0 +1,242 @@
+	switch name {
+	case gitproviders.GitProviderGitHub:
+		return NewGithubDeviceFlowHandler(http.DefaultClient), nil
+
+	}
+
+	return nil, fmt.Errorf("unsupported auth provider \"%s\"", name)
+}
+
+type AuthService struct {
+	logger     logger.Logger
+	fluxClient flux.Flux
+	// Note that this is a k8s go-client, NOT a wego kube.Kube interface.
+	// That interface wasn't providing any valuable abstraction for this service.
+	k8sClient   client.Client
+	gitProvider gitproviders.GitProvider
+	gitClient   git.Git
+}
+
+// NewAuthService constructs an auth service for doing git and GitProvider operations with authenticated clients.
+func NewAuthService(fluxClient flux.Flux, k8sClient client.Client, providerName gitproviders.GitProviderName, l logger.Logger, token string) (*AuthService, error) {
+	provider, err := gitproviders.New(gitproviders.Config{Provider: providerName, Token: token})
+	if err != nil {
+		return nil, fmt.Errorf("error creating git provider client: %w", err)
+	}
+	return &AuthService{
+		logger:      l,
+		fluxClient:  fluxClient,
+		k8sClient:   k8sClient,
+		gitProvider: provider,
+		// We do not have a gitClient yet. We need to get or create deploy keys first.
+		// This will get populated by the SetupDeployKeys call.
+		// Listing it here for the sake of readability (it would still be nil without this entry).
+		gitClient: nil,
+	}, nil
+}
+
+// SetupDeployKey instruments the AuthService with the neccesary data to provide a git client that uses deploy keys.
+// The user MUST call this function before calling .GitClient(), else an error will be returned.
+// The work of setting up deploy keys was NOT done in the constructor for the sake being explicit about what is happening.
+func (a *AuthService) SetupDeployKey(ctx context.Context, name types.NamespacedName, targetName string, repo gitproviders.NormalizedRepoURL) error {
+	owner := repo.Owner()
+	repoName := repo.RepositoryName()
+	accountType, err := a.gitProvider.GetAccountType(repo.Owner())
+	if err != nil {
+		return fmt.Errorf("error getting account type: %w", err)
+	}
+
+	repoInfo, err := a.gitProvider.GetRepoInfo(accountType, repo.Owner(), repo.RepositoryName())
+	if err != nil {
+		return fmt.Errorf("error getting repo info: %w", err)
+	}
+
+	if repoInfo.Visibility != nil && *repoInfo.Visibility == gitprovider.RepositoryVisibilityPublic {
+		// This is a public repo. We don't need to add deploy keys to it.
+		a.gitClient = git.New(nil)
+		return nil
+	}
+
+	deployKeyExists, err := a.gitProvider.DeployKeyExists(owner, repoName)
+	if err != nil {
+		return fmt.Errorf("failed check for existing deploy key: %w", err)
+	}
+
+	if deployKeyExists {
+		a.logger.Println("Existing deploy key found")
+		secretName := types.NamespacedName{
+			Name:      gitproviders.CreateAppSecretName(targetName, repo),
+			Namespace: name.Namespace,
+		}
+		// The deploy key was found on the Git Provider, fetch it from the cluster.
+		secret, err := a.retreiveDeployKey(ctx, secretName)
```

Comment:
Shadow declaration on `err`.  Consider renaming to `errDeployKey` or something along those lines.

Reply:
`err` has to be nil before this line. Do we need to avoid shadowing `err` everywhere?

Reply:
At my last place we always avoided shadow variables.  Eventually it always would bite us and it was always subtle and took a lot of debugging to find out that was the issue.  After enough times it was a hard no.  You don't have to fix them, but one of them somewhere will get us one day.

Reply:
> err has to be nil before this line. Do we need to avoid shadowing err everywhere?

Also, very valid point.